src/main/java/net/minecraftforge/common/MapGenStructureManager.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import net.minecraft.entity.EnumCreatureType;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.World;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.chunk.ChunkPrimer;
+import net.minecraft.world.gen.structure.MapGenStructure;
+import org.apache.commons.lang3.ArrayUtils;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+
+/**
+ * Manage MapGenStructure for the ChunkGenerators.
+ */
+public class MapGenStructureManager
+{
+    private static final List<String> allStructureNames = Lists.newArrayList();
+    private final Map<String, MapGenStructure> structureMap;
+
+    public MapGenStructureManager(List<MapGenStructure> structures)
+    {
+        ImmutableMap.Builder<String, MapGenStructure> builder = ImmutableMap.builder();
+        for (MapGenStructure s : structures)
+        {
+            String name = s.getStructureName();
+            builder.put(name, s);
+            addStructureName(name);
+        }
+        structureMap = builder.build();
+    }
+
+    /**
+     * Add a structure name to the locate command autocomplete list.
+     * <p>
+     * Is automatically done on construction
+     *
+     * @param structure Name of the structure
+     */
+    protected static void addStructureName(String structure)
+    {
+        if (!allStructureNames.contains(structure))
+        {
+            allStructureNames.add(structure);
+        }
+    }
+
+    /**
+     * @return A new array containing both the given strings as well as the registered structure names
+     */
+    public static String[] appendStructureNames(String[] vanilla)
+    {
+        return ArrayUtils.addAll(vanilla, allStructureNames.toArray(new String[0]));
```

Comment:
should we sort it so any command tab completes, etc. don't just a registration-order jumble?

Reply:
maybe a good idea, but then I would stick with a list and sort the names upon registration

Reply:
Sorting the allStructuresNames list in the MapGenStructureManager constructor now. toArray should keep the order.
This way the vanilla order is not influenced, but the additional ones are not based on registration order

Reply:
Is this what you had in mind?